browser-specific-failures.js

```diff
@@ -0,0 +1,242 @@
+const lib = require('./lib');
+const moment = require('moment');
+
+flags.defineString('from', '2018-07-01', 'Starting date (inclusive)');
+flags.defineString('to', moment().format('YYYY-MM-DD'),
+    'Ending date (exclusive)');
+flags.defineStringList('products', ['chrome', 'firefox', 'safari'],
+    'Browsers to compare. Must match the products used on wpt.fyi');
+flags.defineString('output', null,
+    'Output CSV file to write to. Defaults to ' +
+    '{stable, experimental}-browser-specific-failures.csv');
+flags.defineBoolean('experimental', false,
+    'Calculate metrics for experimental runs.');
+flags.parse();
+
+// See documentation of advanceDateToSkipBadDataIfNecessary. These ranges are
+// inclusive, exclusive.
+const STABLE_BAD_RANGES = [
+  // This was some form of Safari outage, undiagnosed but a clear erroneous
+  // spike in failure rates.
+  [moment('2019-02-06'), moment('2019-03-04')],
+  // This was a safaridriver outage, resolved by
+  // https://github.com/web-platform-tests/wpt/pull/18585
+  [moment('2019-06-27'), moment('2019-08-23')],
+];
+const EXPERIMENTAL_BAD_RANGES = [
+  // This was a safaridriver outage, resolved by
+  // https://github.com/web-platform-tests/wpt/pull/18585
+  [moment('2019-06-27'), moment('2019-08-23')],
+];
+
+// There have been periods where results cannot be considered valid and
+// contribute noise to the metrics. Given a date, this function advances it as
+// necessary to avoid bad data.
+//
+// TODO(smcgruer): Take into account --products being used.
+function advanceDateToSkipBadDataIfNecessary(date, experimental) {
+  const ranges = experimental ? EXPERIMENTAL_BAD_RANGES : STABLE_BAD_RANGES;
+  for (const range of ranges) {
+    if (date >= range[0] && date < range[1]) {
+      console.log(`Skipping from ${date.format('YYYY-MM-DD')} to ` +
+          `${range[1].format('YYYY-MM-DD')} due to bad data`);
+      return range[1];
+    }
+  }
+  return date;
+}
+
+const RUNS_URI = 'https://wpt.fyi/api/runs?aligned=true&max-count=1';
+
+// Fetches aligned runs from the wpt.fyi server, between the |from| and |to|
+// dates. If |experimental| is true fetch experimental runs, else stable runs.
+// Returns a map of date to list of runs for that date (one per product)
+//
+// TODO: Known problem: there are periods of time, mostly mid-late 2018, where
+// we ran both Safari 11.1 and 12.1, and the results are massively different.
+// We should fetch multiple runs for each browser and have upgrade logic.
+async function fetchAlignedRunsFromServer(products, from, to, experimental) {
+  const label = experimental ? 'experimental' : 'stable';
+  let params = `&label=master&label=${label}`;
+  for (const product of products) {
+    params += `&product=${product}`;
+  }
+  const runsUri = `${RUNS_URI}${params}`;
+
+  console.log(`Fetching aligned runs from ${from.format('YYYY-MM-DD')} ` +
+      `to ${to.format('YYYY-MM-DD')}`);
+
+  let cachedCount = 0;
+  const before = moment();
```

Comment:
I'm going to punt on this one until a future PR, since making things into closures means pulling out variables that have to be referenced later so it's a bit messy.